lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -193,8 +223,317 @@ struct LowerClassesPass : public LowerClassesBase<LowerClassesPass> {
+    auto owningModule = owningModules.lookup(id);
+    if (!owningModule) {
+      PathInfoTableEntry entry;
+      // This entry is used for checking id uniquness.
+      entry.op = op;
+      entry.id = id;
+      entries.push_back(entry);
+      return true;
+    }
+
+    // Copy the middle part from the annotation's NLA.
+    if (hierPathOp) {
+      // Copy the old path, dropping the module name.
+      auto oldPath = hierPathOp.getNamepath().getValue();
+      llvm::append_range(path, llvm::reverse(oldPath.drop_back()));
+
+      // Set the moduleName based on the hierarchical path. If the
+      // owningModule is in the hierarichal path, set the moduleName to the
+      // owning module. Otherwise use the top of the hierarchical path.
+      bool pathContainsOwningModule =
+          llvm::any_of(oldPath, [&](auto pathFragment) {
+            return llvm::TypeSwitch<Attribute, bool>(pathFragment)
+                .Case([&](hw::InnerRefAttr innerRef) {
+                  return innerRef.getModule() ==
+                         owningModule.getModuleNameAttr();
+                })
+                .Case([&](FlatSymbolRefAttr symRef) {
+                  return symRef.getAttr() == owningModule.getModuleNameAttr();
+                })
+                .Default([](auto attr) { return false; });
+          });
+      if (pathContainsOwningModule) {
+        moduleName = owningModule.getModuleNameAttr();
+      } else {
+        moduleName = cast<hw::InnerRefAttr>(oldPath.front()).getModule();
+      }
+    }
+
+    // Copy the leading part of the hierarchical path from the owning module
+    // to the start of the annotation's NLA.
+    auto needsAltBasePath =
+        getOrComputeNeedsAltBasePath(op->getLoc(), moduleName, owningModule);
+    if (failed(needsAltBasePath)) {
+      error = true;
+      return false;
+    }
+
+    // Create the HierPathOp.
+    std::reverse(path.begin(), path.end());
+    auto pathAttr = ArrayAttr::get(op->getContext(), path);
+
+    // If we need an alternative base path, save the top module from the
+    // path. We will plumb in the basepath from this module.
+    StringAttr altBasePathModule;
+    if (*needsAltBasePath) {
+      altBasePathModule =
+          TypeSwitch<Attribute, StringAttr>(path.front())
+              .Case<FlatSymbolRefAttr>([](auto a) { return a.getAttr(); })
+              .Case<hw::InnerRefAttr>([](auto a) { return a.getModule(); });
+
+      altBasePathRoots.insert(altBasePathModule);
+    }
+
+    // Record the path operation associated with the path op.
+    entries.push_back({op, id, altBasePathModule, pathAttr});
+
+    // Remove this annotation from the operation.
+    return true;
+  });
+  if (error)
+    return failure();
+
+  if (isa<PortAnnoTarget>(target))
+    // Port annotations are batch-updated afterwards.
+    portAnnotations.push_back(annotations.getArrayAttr());
+  else
+    // Otherwise it's fine to update immediately.
+    target.setAnnotations(annotations);
```

Comment:
` target.setAnnotations(annotations);` could be made slightly faster with ` cast<OpAnnoTarget>(target)setAnnotations(annotations);`

I don't love something about the way this sometimes mutates an internal structure and other times just sets the annotations. I think I would be happier if this function always appended to an annotation array, and op modifications happened outside this function, and then `portAnnotations` doesn't need to be a member anymore.  It would be even nicer if this whole function didn't care about the target type at all - but we need to know when we create the `targetSym`. WDYT?

Reply:
I changed processPathTrackers to return `FailureOr<AnnotationSet>` and mutate annotations from callers. Does that look good to you?